system/security/securesocket/securesocket.cpp

```diff
@@ -1641,3 +1641,71 @@ SECURESOCKET_API int signCertificate(const char* csr, const char* ca_certificate
 }
 
 }
+
+ISmartSocketException *createSecureSmartSocketException(int errorCode, const char *msg)
+{
+    return new SmartSocketException(errorCode, msg);
+}
+
+class CSecureSmartSocketFactory : public CSmartSocketFactory
+{
+public:
+    Owned<ISecureSocketContext> secureContext;
+
+    CSecureSmartSocketFactory(const char *_socklist, bool _retry, unsigned _retryInterval, unsigned _dnsInterval) : CSmartSocketFactory(_socklist, _retry, _retryInterval, _dnsInterval)
+    {
+        secureContext.setown(createSecureSocketContext(ClientSocket));
+    }
+
+    ISmartSocket *connect_timeout(unsigned timeoutms)
+    {
+        SmartSocketEndpoint *ss = nextSmartEndpoint();
+        if (!ss)
+            throw createSecureSmartSocketException(0, "securesmartsocket failed to get nextEndpoint");
+
+        ISecureSocket *ssock = nullptr;
+        ISocket *sock = nullptr;
+        SocketEndpoint ep;
+        try
+        {
+            {
+                synchronized block(lock);
+                ss->checkHost(dnsInterval);
+                ep = ss->ep;
+            }
+            if (timeoutms)
+                sock = ISocket::connect_timeout(ep, timeoutms);
+            else
+                sock = ISocket::connect(ep);
+
+            ssock = secureContext->createSecureSocket(sock);
+
+            int status = ssock->secure_connect();
+            if (status < 0)
+            {
+                // secure_connect may also DBGLOG() errors ...
+                ssock->Release();
+                secureContext->Release();
+                throw MakeStringException(1, "SSL connect fail");
+            }
```

Comment:
I think apart from a few lines above, this is a same code as in CSmartSocketFactory.
It would be better to make a common base method, that created and returned the base socket, then call that here and in CSmartSocketFactory::connect_timeout